source/h1_connection.c

```diff
@@ -587,17 +587,31 @@ static void s_stream_complete(struct aws_h1_stream *stream, int error_code) {
             AWS_BYTE_CURSOR_PRI(stream->base.server_data->request_method_str));
     }
 
+    /* If connection must shut down, do it BEFORE invoking stream-complete callback.
+     * That way, if aws_http_connection_is_open() is called from stream-complete callback, it returns false. */
+    if (!original_error_code) {
+        if (error_code) {
+            s_shutdown_due_to_error(connection, error_code);
+
+        } else if (stream->is_final_stream) {
+            /* TODO: Also close servers. RFC-7230 section 6.6 warns that this is tricky. */
+            if (connection->base.client_data) {
+                AWS_LOGF_TRACE(
+                    AWS_LS_HTTP_CONNECTION,
+                    "id=%p: Closing connection due to completion of final stream.",
+                    (void *)&connection->base);
+
+                s_shutdown_due_to_error(connection, AWS_ERROR_HTTP_CONNECTION_CLOSED);
```

Comment:
This isn't necessarily an error right? We don't want the channel shutting down on an error. It's not a protocol error, nothing went wrong. We just want the connection to close. The user should look at the response code to figure out whether or not it's an error condition right?

Reply:
I wasn't sure what error code to use for a closed connection that wasn't initiated by the user calling `aws_http_connection_close()`.

We talked, and we'll consider this to be a type of clean shutdown.